server/request_handler.go

```diff
@@ -0,0 +1,173 @@
+package server
+
+import (
+	"crypto/ecdsa"
+	"encoding/json"
+	"fmt"
+	"github.com/flashbots/rpc-endpoint/types"
+	"github.com/flashbots/rpc-endpoint/utils"
+	"github.com/google/uuid"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type RequestHandler interface {
+	ProcessRequest() *types.JsonRpcResponse
+}
+
+// RPC request handler for a single/ batch JSON-RPC request
+type RpcRequestHandler struct {
+	respw *http.ResponseWriter
+	req   *http.Request
+
+	uid             string
+	timeStarted     time.Time
+	defaultProxyUrl string
+	relaySigningKey *ecdsa.PrivateKey
+
+	handleBatch bool // If set, batch request will be handled
+}
+
+func NewRpcRequestHandler(respw *http.ResponseWriter, req *http.Request, proxyUrl string, relaySigningKey *ecdsa.PrivateKey) *RpcRequestHandler {
+	return &RpcRequestHandler{
+		respw:           respw,
+		req:             req,
+		uid:             uuid.New().String(),
+		timeStarted:     Now(),
+		defaultProxyUrl: proxyUrl,
+		relaySigningKey: relaySigningKey,
+	}
+}
+
+func (r *RpcRequestHandler) log(format string, v ...interface{}) {
+	prefix := fmt.Sprintf("[%s] ", r.uid)
+	log.Printf(prefix+format, v...)
+}
+
+func (r *RpcRequestHandler) logError(format string, v ...interface{}) {
+	prefix := fmt.Sprintf("[%s] ERROR: ", r.uid)
+	log.Printf(prefix+format, v...)
+}
+
+func (r *RpcRequestHandler) process() {
+	var err error
+
+	// At end of request, log the time it needed
+	defer func() {
+		timeRequestNeeded := time.Since(r.timeStarted)
+		r.log("request took %.6f sec", timeRequestNeeded.Seconds())
+	}()
+
+	ip := utils.GetIP(r.req)
+	origin := r.req.Header.Get("Origin")
+	r.log("POST request from ip: %s - origin: %s", ip, origin)
+
+	if IsBlacklisted(ip) {
+		r.log("Blocked IP: %s", ip)
+		r._writeHeaderStatus(http.StatusUnauthorized)
+		return
+	}
+
+	// If users specify a proxy url in their rpc endpoint they can have their requests proxied to that endpoint instead of Infura
+	// e.g. https://rpc.flashbots.net?url=http://RPC-ENDPOINT.COM
+	customProxyUrl, ok := r.req.URL.Query()["url"]
+	if ok && len(customProxyUrl[0]) > 1 {
+		r.defaultProxyUrl = customProxyUrl[0]
+		r.log("Using custom url: %s", r.defaultProxyUrl)
+	}
+
+	// Decode request JSON RPC
+	defer r.req.Body.Close()
+	body, err := ioutil.ReadAll(r.req.Body)
+	if err != nil {
+		r.logError("failed to read request body: %v", err)
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	if len(body) == 0 {
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	// Parse JSON RPC payload
+	var data interface{}
+	if data, err = r.UnmarshalJSON(body); err != nil {
+		r.log("Parse payload %v", err)
+		r._writeHeaderStatus(http.StatusBadRequest)
+		return
+	}
+
+	// Handle batch request
+	if r.handleBatch {
+		batchCh := make(chan *types.JsonRpcResponse) // Chan to hold response from each go routine
+		wg := new(sync.WaitGroup)                    // Init Wait group
```

Comment:
I think using the scatter-gather pattern here would be more elegant: https://peter.bourgon.org/go-for-industrial-programming/?s=09#scatter-gather

Reply:
Used scatter gather patter as you have shown. Thank you for the pointer :-)